In [1]:
# link :
# https://projecteuler.net/archives

# Problem 1 (Multiples of 3 and 5)

If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of3 and 5 below 1000, not up to and equal 1000.

In [2]:
# ref:https://math.stackexchange.com/questions/9259/find-the-sum-of-all-the-multiples-of-3-or-5-below-1000

In [3]:
multi_3 = []
multi_5 = []
multi_3and5 = []
for i in range(1,1000,1):
    if i % 3 ==0:
        multi_3.append(i)    
    if i % 5 ==0:
        multi_5.append(i)
    if i % 3 == 0 and i % 5 == 0:
        multi_3and5.append(i)

print (multi_3)
print (multi_5)
print(multi_3and5)
print (sum(multi_3)+sum(multi_5)-sum(multi_3and5)) # subtracting to avoid double counting #answer

[3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99, 102, 105, 108, 111, 114, 117, 120, 123, 126, 129, 132, 135, 138, 141, 144, 147, 150, 153, 156, 159, 162, 165, 168, 171, 174, 177, 180, 183, 186, 189, 192, 195, 198, 201, 204, 207, 210, 213, 216, 219, 222, 225, 228, 231, 234, 237, 240, 243, 246, 249, 252, 255, 258, 261, 264, 267, 270, 273, 276, 279, 282, 285, 288, 291, 294, 297, 300, 303, 306, 309, 312, 315, 318, 321, 324, 327, 330, 333, 336, 339, 342, 345, 348, 351, 354, 357, 360, 363, 366, 369, 372, 375, 378, 381, 384, 387, 390, 393, 396, 399, 402, 405, 408, 411, 414, 417, 420, 423, 426, 429, 432, 435, 438, 441, 444, 447, 450, 453, 456, 459, 462, 465, 468, 471, 474, 477, 480, 483, 486, 489, 492, 495, 498, 501, 504, 507, 510, 513, 516, 519, 522, 525, 528, 531, 534, 537, 540, 543, 546, 549, 552, 555, 558, 561, 564, 567, 570, 573, 576, 579, 582, 585, 588, 591, 594, 597, 600, 603, 606, 609, 612, 615, 618, 621, 

# Problem 2  (Even Fibonacci numbers)

Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [4]:
a = [1,2]
i=0
while (a[i]+a[i+1]<=4000000):
    a.append(a[i]+a[i+1])
    i=i+1
print(a)

b =[]
for i in range(len(a)):
    if a[i]%2 ==0:
        b.append(a[i])
print(b)        
print(sum(b)) #answer

[1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578]
[2, 8, 34, 144, 610, 2584, 10946, 46368, 196418, 832040, 3524578]
4613732


# Problem 3 (Largest prime factor)

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

In [5]:
# ref: https://stackoverflow.com/questions/15347174/python-finding-prime-factors

In [6]:
n = int( 600851475143 )
factors = []
for i in range(2,n+1): #1 is neither prime nor composite
    while n % i == 0:
        n = n/i # to get repeated prime factor
        factors.append(i)
    if n == 1: 
        break
print (factors)
print (max(factors)) # answer

[71, 839, 1471, 6857]
6857


In [16]:
n = int( 600851475143 )
factors = []
for i in range(2,int(np.sqrt( 600851475143))+1): # 1 is neither prime nor composite
                                                 # This range is based on the fact that a composite number must have a factor less than the square root of that number.
    while n % i == 0:
        n = n/i # to get repeated prime factor
        factors.append(i)
    if n == 1: 
        break
print (factors)
print (max(factors)) # answer

[71, 839, 1471, 6857]
6857


# Problem 4 (Largest palindrome product)

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

In [7]:
list_of_palindrome = []
for num in range(999,101,-1):
    i = 0
    while(num-i>=100):
        product_of_3_digit_num = num*(num-i)
        if str(product_of_3_digit_num)==(str(product_of_3_digit_num))[::-1]:
#             print(product_of_3_digit_num)
            list_of_palindrome.append(product_of_3_digit_num)
#             print(list_of_palindrome)
            i=i+1
        else:
            i=i+1
print(list_of_palindrome)
print(max(list_of_palindrome)) #answer

[580085, 514415, 906609, 119911, 282282, 141141, 853358, 650056, 601106, 592295, 543345, 485584, 436634, 378873, 329923, 408804, 204402, 623326, 525525, 299992, 469964, 886688, 804408, 698896, 631136, 616616, 443344, 428824, 270072, 255552, 828828, 414414, 487784, 180081, 888888, 666666, 444444, 222222, 348843, 770077, 653356, 855558, 807708, 696696, 648846, 531135, 489984, 372273, 324423, 165561, 117711, 577775, 723327, 649946, 272272, 219912, 292292, 840048, 821128, 802208, 630036, 611116, 420024, 401104, 210012, 238832, 227722, 595595, 589985, 560065, 554455, 548845, 513315, 507705, 564465, 174471, 861168, 809908, 470074, 489984, 693396, 672276, 579975, 828828, 801108, 696696, 636636, 444444, 279972, 252252, 219912, 476674, 780087, 393393, 262262, 131131, 443344, 513315, 824428, 412214, 793397, 747747, 650056, 604406, 588885, 491194, 445544, 302203, 286682, 143341, 551155, 99299, 819918, 729927, 639936, 549945, 459954, 369963, 279972, 189981, 657756, 507705, 407704, 168861, 661166, 

# Problem 5 (Smallest multiple)

2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

In [8]:
#ref
import itertools
list = [[1, 2, 0], [1, 2, 1], [7, 45, 0]]
for x in itertools.takewhile(lambda x: x[2] == 0, list):
    print (x)
print ('------------------')
for x in filter(lambda x: x[2] == 0, list):
    print (x)

[1, 2, 0]
------------------
[1, 2, 0]
[7, 45, 0]


In [9]:
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
# USING FOR LOOP
import numpy as np
l = np.linspace(1,10,10)
# print (l)
result =[]

for i in range(1,5000,1):
    a=[]
    for j in l:
        if i%j==0:
            a.append('True')
    if len(a)==10:
        print(a)
        print(i)
        result.append(i)  

print ('Result:'+str(result))

['True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True']
2520
Result:[2520]


In [10]:
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
# USING WHILE LOOP
import numpy as np
l = np.linspace(1,10,10)
# print (l)
result =[]
i=1
while (True):
    a=[]
    for j in l:
        if i%j==0:
            a.append('True')
    if len(a)==10:
        print(a)
        print(i)
        result.append(i)
        break
    i=i+1

print ('Result:'+str(result))

['True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True']
2520
Result:[2520]


In [11]:
# Checking every number taking lot of time. So below idea is needed 
# (the number shd be of 20 table, no need to check all numbers)
# ******* ref: https://www.mathblog.dk/project-euler-problem-5/
# We can rather easily speed it up, since we would need the number to be divisible by 20, 
# we don’t need to check every number, but we can limit it to search in increments of 20. 


# USING WHILE LOOP
import numpy as np
l = np.linspace(1,20,20)
result =[]
i=20
while (True):
    a=[]
    for j in l:
        if i%j==0:
            a.append('True')
    if len(a)==len(l):
        print(a)
        print(i)
        result.append(i)
        break
    i=i+20

print ('Result:'+str(result))  
print ('Answer:'+str(min(result))) #answer

['True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True', 'True']
232792560
Result:[232792560]
Answer:232792560


# Problem 6 (Sum square difference)

The sum of the squares of the first ten natural numbers is,

1^2+2^2+...+10^2=385
The square of the sum of the first ten natural numbers is,

(1+2+...+10)^2=552=3025
Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is 3025−385=2640.

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.


In [14]:
l = np.linspace(1,100,100)
a=[]
for j in l:
    a.append(j**2)
sum_1 = sum(a)
sum_2 = (sum(l))**2
print (sum_2-sum_1) #answer

25164150.0


# Problem 7 (10001st prime)

By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10001st prime number?

The function is_prime_number() returns False if the number supplied is less than 2 and if the number is equally divisible with some other number different than 1 and itself. If none of the previous conditions apply the function will return True.

In [29]:
n = 10001  #What is nth prime number

def is_prime_number(x):
    if x >= 2:
        #for i in range(2,x): #different than 1 and itself : so starting range with 2 and ending with x i.e, [2,3...x-1]
        for i in range(2,int(np.sqrt(x))+1): # This range is based on the fact that a composite number must have a factor less than the square root of that number.
            if x % i == 0:
                return False
    else:
        return False
    return True 



prime_numbers = []
i=0
while(True):
    if is_prime_number(i):
            prime_numbers.append(i)
    i=i+1
    if len(prime_numbers)==n:
        break
          
# print(prime_numbers)
print(prime_numbers[-1]) #answer

104743


# Problem 8 (Largest product in a series)

The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

73167176531330624919225119674426574742355349194934
96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450

Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?


In [49]:
# https://www.geeksforgeeks.org/reduce-in-python/
# https://nekedome-project-euler.readthedocs.io/en/latest/solutions/8.html

In [52]:
from functools import reduce
from operator import mul


def solve():
    """ Compute the answer to Project Euler's problem #8 """

    # Build a list of the individual digits as integer objects
    series = """
        73167176531330624919225119674426574742355349194934
        96983520312774506326239578318016984801869478851843
        85861560789112949495459501737958331952853208805511
        12540698747158523863050715693290963295227443043557
        66896648950445244523161731856403098711121722383113
        62229893423380308135336276614282806444486645238749
        30358907296290491560440772390713810515859307960866
        70172427121883998797908792274921901699720888093776
        65727333001053367881220235421809751254540594752243
        52584907711670556013604839586446706324415722155397
        53697817977846174064955149290862569321978468622482
        83972241375657056057490261407972968652414535100474
        82166370484403199890008895243450658541227588666881
        16427171479924442928230863465674813919123162824586
        17866458359124566529476545682848912883142607690042
        24219022671055626321111109370544217506941658960408
        07198403850962455444362981230987879927244284909188
        84580156166097919133875499200524063689912560717606
        05886116467109405077541002256983155200055935729725
        71636269561882670428252483600823257530420752963450
    """
    print (type(series), len(series))
    series = series.replace(" ", "").replace("\n", "")
    print (type(series), len(series))
    integers = [int(character) for character in series]
    print (type(integers), len(integers))

    # Perform the search through all overlapping m-long subsets
    n = len(integers)
    m = 13
    answer = 0
    for i in range(n - m + 1):
        subset = integers[i:i+m]
        product = reduce(mul, subset, 1)
        answer = max(answer, product)
    return answer

print(solve())

<class 'str'> 1185
<class 'str'> 1000
<class 'list'> 1000
23514624000


# Problem 9 (Special Pythagorean triplet)

A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

a^2 + b^2 = c^2
For example, 3^2 + 4^2 = 9 + 16 = 25 = 52.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.

In [14]:
import itertools
a = [i for i in range(1,1000,1)]
b = [i for i in range(1,1000,1)]
l = list(itertools.product(a, b))
# print (l, type(l), l[0][0],l[1][0], len(l))

for i in range(len(l)):
    C = l[i][0]**2 + l[i][1]**2
    c = C**0.5
    if c%2 == 0 or (c+1)%2 == 0: # to check if sqrt value is a natural number
        if l[i][0] + l[i][1]+c==1000:
            print (l[i][0],l[i][1],c)
            d = l[i][0]*l[i][1]*c
            break
        
print ('Result:'+str(d)) # answer    

200 375 425.0
Result:31875000.0


# Problem 10 (Summation of primes)

The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.

In [18]:
import numpy as np
n = 2000000  

def is_prime_number(x):
    if x >= 2:
        #for i in range(2,x): #different than 1 and itself : so starting range with 2 and ending with x i.e, [2,3...x-1]
        for i in range(2,int(np.sqrt(x))+1): # This range is based on the fact that a composite number must have a factor less than the square root of that number.
            if x % i == 0:
                return False
    else:
        return False
    return True 



prime_numbers = []
i=0
while(True):
    if is_prime_number(i):
            prime_numbers.append(i)
    if i>=n:
        break
    i=i+1
          
print(sum(prime_numbers))

142913828922
